src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp

```diff
@@ -28,249 +29,823 @@
+               _leftmosts[int(ShenandoahFreeSetPartitionId::Collector)],
+               _rightmosts[int(ShenandoahFreeSetPartitionId::Collector)]);
+  log_info(gc)("Empty Mutator range [" SSIZE_FORMAT ", " SSIZE_FORMAT
+               "], Empty Collector range [" SSIZE_FORMAT ", " SSIZE_FORMAT "]",
+               _leftmosts_empty[int(ShenandoahFreeSetPartitionId::Mutator)],
+               _rightmosts_empty[int(ShenandoahFreeSetPartitionId::Mutator)],
+               _leftmosts_empty[int(ShenandoahFreeSetPartitionId::Collector)],
+               _rightmosts_empty[int(ShenandoahFreeSetPartitionId::Collector)]);
+
+  log_info(gc)("%6s: %18s %18s %18s", "index", "Mutator Bits", "Collector Bits", "NotFree Bits");
+  dump_bitmap_range(0, _max-1);
+}
+
+void ShenandoahRegionPartitions::dump_bitmap_range(idx_t start_idx, idx_t end_idx) const {
+  assert((start_idx >= 0) && (start_idx < (idx_t) _max), "precondition");
+  assert((end_idx >= 0) && (end_idx < (idx_t) _max), "precondition");
+  idx_t aligned_start = _membership[int(ShenandoahFreeSetPartitionId::Mutator)].aligned_index(start_idx);
+  idx_t aligned_end = _membership[int(ShenandoahFreeSetPartitionId::Mutator)].aligned_index(end_idx);
+  idx_t alignment = _membership[int(ShenandoahFreeSetPartitionId::Mutator)].alignment();
+  while (aligned_start <= aligned_end) {
+    dump_bitmap_row(aligned_start);
+    aligned_start += alignment;
+  }
+}
+
+void ShenandoahRegionPartitions::dump_bitmap_row(idx_t idx) const {
+  assert((idx >= 0) && (idx < (idx_t) _max), "precondition");
+  idx_t aligned_idx = _membership[int(ShenandoahFreeSetPartitionId::Mutator)].aligned_index(idx);
+  uintx mutator_bits = _membership[int(ShenandoahFreeSetPartitionId::Mutator)].bits_at(aligned_idx);
+  uintx collector_bits = _membership[int(ShenandoahFreeSetPartitionId::Collector)].bits_at(aligned_idx);
+  uintx free_bits = mutator_bits | collector_bits;
+  uintx notfree_bits =  ~free_bits;
+  log_info(gc)(SSIZE_FORMAT_W(6) ": " SIZE_FORMAT_X_0 " 0x" SIZE_FORMAT_X_0 " 0x" SIZE_FORMAT_X_0,
+               aligned_idx, mutator_bits, collector_bits, notfree_bits);
+}
+
+ShenandoahRegionPartitions::ShenandoahRegionPartitions(size_t max_regions, ShenandoahFreeSet* free_set) :
+    _max(max_regions),
+    _region_size_bytes(ShenandoahHeapRegion::region_size_bytes()),
+    _free_set(free_set),
+    _membership{ ShenandoahSimpleBitMap(max_regions), ShenandoahSimpleBitMap(max_regions) }
 {
-  clear_internal();
+  make_all_regions_unavailable();
 }
 
-void ShenandoahFreeSet::increase_used(size_t num_bytes) {
-  shenandoah_assert_heaplocked();
-  _used += num_bytes;
+inline bool ShenandoahFreeSet::can_allocate_from(ShenandoahHeapRegion *r) const {
+  return r->is_empty() || (r->is_trash() && !_heap->is_concurrent_weak_root_in_progress());
+}
+
+inline bool ShenandoahFreeSet::can_allocate_from(size_t idx) const {
+  ShenandoahHeapRegion* r = _heap->get_region(idx);
+  return can_allocate_from(r);
+}
```

Comment:
Some of this abstraction leakage can be avoided by making

"can_allocate_from(size_t idx)" a method on the `_heap` object, and calling that, letting the heap do the test:

``` 
bool ShenandoahHeap::can_allocate_from(size_t idx) {
  ShenandoahHeapRegion* r = get_region(idx);
  return r->is_empty() || (r->is_trash() && !is_concurrent_weak_root_in_progress());
}
```

with ShFreeSet merely delegating to its `_heap`.

Probably similar pattern for `alloc_capacity` method below.

(I'll keep leaving single comments, so I don't slow down the review/feedback across potential interruptions. I don't expect there will be many, but still.)

Reply:
I'm not sure I understand the rationale for these changes.  In my mind, the "notions" of can_allocate_from() and alloc_capacity() as used here within ShenandoahFreeSet are not universal.  These notions are specific to what the free set understands about the life-cycle of regions.  I think moving this functionality into heap actually leaks more encapsulation, in that nobody else cares to use these novel interpretations of can_allocate_from() and alloc_capacity().

Maybe you can elaborate?

Reply:
I was looking at the structural effect of moving the methods there and failed to see that these were policies encoded in the free set abstraction rather than policies of (or characteristics of) ShenandoahHeap _used_ by the free set (as I was thinking).

No change is needed in that case. Sorry for my confusion.